lib/galaxy/managers/base.py

```diff
@@ -999,7 +999,7 @@ class ModelFilterParser(HasAModelManager):
     # (as the model informs how the filter params are parsed)
     # I have no great idea where this 'belongs', so it's here for now
 
-    model_class: Type[model._HasTable]
+    model_class: Type[Union[model.Base, model.DatasetInstance]]
```

Comment:
`DatasetInstance` doesn't look right here. If you do need to be more precise consider parameterizing this with a generic ?

Reply:
We need a union here only because hda and ldda are not mapped declaratively. `model_class` is the type for all SA models. All models *but these two* are mapped declaratively, which gives them a `metadata` instance variable. The hda and ldda models also have a `metadata` variable, which is why they can't be mapped declaratively (they will be eventually). Now, all declaratively mapped models inherit from `DeclarativeBase`. So the closest way to describe the type of *all our models* is the union of `DeclarativeBase` *and* the 2 non-declaratively mapped models - which both subclass `DatasetInstance`. This is tighter than the previous implementation (`_HasTable`, before that `_BaseModel`) which relied on the parent having a `table` variable.